### ISSUE TITLE
MSN: Masked Siamese Networks for Label-Efficient Learning

### ISSUE THREAD

**Issue by masc-it** [lightly-ai/lightly#878]
Hello, I would like to share this new interesting SSL method proposed by Meta that achieves SOTA results and shows interesting generalization properties.

It is based on ViT and I think it could be worth to integrate it in this codebase. Also, the official implementation is available :)

[MSN Paper](https://arxiv.org/abs/2204.07141), [MSN GIthub code](https://github.com/facebookresearch/msn)

**Comment by IgorSusmelj** [comment]
@masc-it, this is a great idea. I haven't read the paper yet and will do it for our paper session. We will discuss the paper in our upcoming session this Friday 1pm CET. Feel free to join our discord channel: https://discord.gg/xvNJW94

After the session we hopefully have a better understanding of how the model could be best integrated. If you already had some thoughts feel free to share them :)

**Comment by philippmwirth** [comment]
Hey @masc-it this should work almost out of the box. We have the necessary tools for momentum updates here:
- https://github.com/lightly-ai/lightly/blob/c1c4ee2b2de8e3fa2e831cf142d4864e80c0b448/lightly/models/utils.py#L159
- https://github.com/lightly-ai/lightly/blob/c1c4ee2b2de8e3fa2e831cf142d4864e80c0b448/lightly/models/utils.py#L186

For masking and ViTs we can use the MAE implementation
- https://github.com/lightly-ai/lightly/blob/c1c4ee2b2de8e3fa2e831cf142d4864e80c0b448/lightly/models/modules/masked_autoencoder.py#L13

The only thing missing then would be the implementation of the loss function which looks rather straightforward. So I'd say we can go ahead and add this to our framework 👍 

@guarin and @MalteEbner what are your thoughts on this?

**Comment by philippmwirth** [comment]
I quickly put this example together:

```python
# Note: The model and training settings do not follow the reference settings
# from the paper. The settings are chosen such that the example can easily be
# run on a small dataset with a single GPU.

import copy
import torch
from torch import nn
import torchvision
import pytorch_lightning as pl

from lightly.data import LightlyDataset
from lightly.data.collate import MAECollateFunction
from lightly.models import utils
from lightly.models.modules import masked_autoencoder
from lightly.models.utils import deactivate_requires_grad
from lightly.models.utils import update_momentum


class MSN(pl.LightningModule):
    def __init__(self):
        super().__init__()
        
        decoder_dim = 512
        vit = torchvision.models.vit_b_32(pretrained=False)
        self.mask_ratio = 0.5
        self.patch_size = vit.patch_size
        self.sequence_length = vit.seq_length
        self.mask_token = nn.Parameter(torch.zeros(1, 1, decoder_dim))
        self.backbone = masked_autoencoder.MAEBackbone.from_vit(vit)
        self.momentum_backbone = copy.deepcopy(self.backbone)
        deactivate_requires_grad(self.momentum_backbone)
        self.criterion = # TODO: implement the loss function

    def training_step(self, batch, batch_idx):

        # momentum update
        update_momentum(self.backbone, self.momentum_backbone, 0.99)

        (anchor_images, target_images), _, _ = batch
        
        batch_size = images.shape[0]
        idx_keep, _ = utils.random_token_mask(
            size=(batch_size, self.sequence_length),
            mask_ratio=self.mask_ratio,
            device=images.device,
        )

        anchor_encoded = self.backbone(anchor_images, idx_keep)
        target_encoded = self.backbone(target_images)
        
        loss = self.criterion(anchor_encoded, target_encoded)

        return loss

    def configure_optimizers(self):
        optim = torch.optim.AdamW(self.parameters(), lr=1.5e-4)
        return optim


model = MSN()

# we ignore object detection annotations by setting target_transform to return 0
pascal_voc = torchvision.datasets.VOCDetection(
    "datasets/pascal_voc", download=True, target_transform=lambda t: 0
)
dataset = LightlyDataset.from_torch_dataset(pascal_voc)
# or create a dataset from a folder containing images or videos:
# dataset = LightlyDataset("path/to/folder")

collate_fn = # TODO: implement MSNCollateFunction

dataloader = torch.utils.data.DataLoader(
    dataset,
    batch_size=256,
    collate_fn=collate_fn,
    shuffle=True,
    drop_last=True,
    num_workers=8,
)

gpus = 1 if torch.cuda.is_available() else 0

trainer = pl.Trainer(max_epochs=10, gpus=gpus)
trainer.fit(model=model, train_dataloaders=dataloader)
```

There are some details I left out here (like the linear scaling of the momentum) but all in all I think it shouldn't be too hard to add it 🙂

**Comment by guarin** [comment]
The proposal looks great!

I noticed two small things that are missing:
* Focal mask augmentation
* Cluster prototypes / assignments

I think both should be relatively easy to add.

**Comment by philippmwirth** [comment]
Thanks for the feedback, @guarin. I went through the paper again and added dummies for focal masks, projection heads, and prototypes. The only thing missing in the proposal above is the "sinkhorn normalization" of the prototypes which doesn't seem to be required for best performance:
![image](https://user-images.githubusercontent.com/65946090/182379444-2c601e58-6c53-4f43-ae0e-81e2f23475ec.png)


So my suggestion would be to follow as similar approach to what's shown above. For this, the following things are currently missing:
- `MSNProjectionHead`: "We train with a 3-layer projection head with output dimension 256 and batch-normalization at the input and hidden layers"
- `MSNCriterion` (or a more generic name maybe)
- `focal_token_mask` (put it in `lightly.models.utils`)

Then, later, we can still write an abstraction of the prototypes with sinkhorn normalization.

**Comment by philippmwirth** [comment]
Quick update, we can't simply use the `SimCLRCollateFunction` as shown above because they use different augmentations for each of the anchor images (hence we need a `MultiViewCollateFunction` instead).

**Comment by philippmwirth** [comment]
Also, correct me if I'm wrong, but I think the `focal_mask` is simply implemented as a crop in the official repo:
https://github.com/facebookresearch/msn/blob/4388dc1eadbe3042b85d3296d41b9b207656e043/src/deit.py#L213

**Comment by japrescott** [comment]
Docs are available: https://docs.lightly.ai/examples/msn.html

**Comment by luzuku** [comment]
@guarin Can you explain why the implementation of the loss does not use `proto_labels`/`support_labels` (which are one-hot vectors) like the original does [here](https://github.com/facebookresearch/msn/blob/4388dc1eadbe3042b85d3296d41b9b207656e043/src/losses.py#L38-L42)? They are initialized [here](https://github.com/facebookresearch/msn/blob/4388dc1eadbe3042b85d3296d41b9b207656e043/src/msn_train.py#L219-L227).

Maybe I overlooked something but to my eyes lightly does this:
```python
softmax(query @ supports.T / temp)
```

and facebookresearch/msn does this:
```python
softmax(query @ supports.T / temp) @ support_labels
```

**Comment by luzuku** [comment]
Ok, so as I understand it, the original implementation has the option for label smoothing the `proto_labels`/ `support_labels` in which case they would be different from the identity matrix, which your implementation does not support. But this option isn't mentioned in the paper at all and the given default config has it [disabled anyways](https://github.com/facebookresearch/msn/blob/4388dc1eadbe3042b85d3296d41b9b207656e043/configs/pretrain/msn_vits16.yaml#L17), so I guess in that sense your implementation does not differ from the original. Would have been nice to mention this in a comment somewhere though.

**Comment by guarin** [comment]
Hi @luzuku, sorry for the late response. But yes we just implemented the default setting and skipped label smoothing. We sometimes skip these special cases as the original code often contains research code that doesn't make it into the final paper.
